### ISSUE TITLE
Server randomly stops the connection when running with eventlet

### ISSUE THREAD

**Issue by egorf** [miguelgrinberg/Flask-SocketIO#337]
Hi!

I am using the latest version of your library, eventlet and the client socket.io on js. From time to time, nearly uncatchable error occurs: the server dies, loses connection to the front end and does not accept new connections after a page refresh. I wish I could provide more debug info, but I don't know how. 

Full lib versions:

```
click==6.6
eventlet==0.19.0
Flask==0.11.1
Flask-SocketIO==2.7.1
greenlet==0.4.10
itsdangerous==0.24
Jinja2==2.8
MarkupSafe==0.23
ptyprocess==0.5.1
python-engineio==1.0.3
python-socketio==1.6.0
six==1.10.0
Werkzeug==0.11.11
wheel==0.24.0
```

**Comment by egorf** [comment]
Here's a traceback a managed to grab. Note that this is after hitting KeyboardInterrupt!

```
^CTraceback (most recent call last):
  File "manage.py", line 14, in <module>
    rtk.stop_base_coordinates_worker()
  File "/home/egor/Projects/ReachView-private/app/rtklib_controller.py", line 100, in stop_base_coordinates_worker
    self.coordinate_averaging_thread.join()
  File "/usr/lib/python2.7/threading.py", line 940, in join
    self.__block.wait()
  File "/usr/lib/python2.7/threading.py", line 340, in wait
    waiter.acquire()
  File "/home/egor/Projects/ReachView-private/venv/local/lib/python2.7/site-packages/eventlet/semaphore.py", line 113, in acquire
    hubs.get_hub().switch()
  File "/home/egor/Projects/ReachView-private/venv/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 294, in switch
    return self.greenlet.switch()
  File "/home/egor/Projects/ReachView-private/venv/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 346, in run
    self.wait(sleep_time)
  File "/home/egor/Projects/ReachView-private/venv/local/lib/python2.7/site-packages/eventlet/hubs/poll.py", line 85, in wait
    presult = self.do_poll(seconds)
  File "/home/egor/Projects/ReachView-private/venv/local/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 62, in do_poll
    return self.poll.poll(seconds)
KeyboardInterrupt
```

**Comment by miguelgrinberg** [comment]
Could you run your server with logging enabled? Just add `engineio_logger=True` to your `SocketIO` constructor. That may give me some clues about your problem.

**Comment by egorf** [comment]
Ok, so I finally managed to recreate it! Still no sane way to do it, except for spamming and waiting.

It seems that the client stops responding, but I'm not sure...If it helps, **pace.js** on the webpage, which is configured **to ignore** WebSocket interactions, does not finish and stays as a blue line on the top. 

The final part of the log looks like this(stderr only):

```
e08b20d2fffa4bd893a787f99d3858cb: Client is gone, closing socket
d674d8ff2a6540659bed1dce7735d4a5: Received packet PING data None
d674d8ff2a6540659bed1dce7735d4a5: Sending packet PONG data None
(29530) accepted ('192.168.1.118', 37583)
522710e7f7fa482495711bef24dc3ab9: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '522710e7f7fa482495711bef24dc3ab9'}
(29530) accepted ('192.168.1.118', 37589)
5ba4b7623c794cb1b280a5bbaeae984e: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '5ba4b7623c794cb1b280a5bbaeae984e'}
0b79674a23a8430bba826d38c5f440b3: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '0b79674a23a8430bba826d38c5f440b3'}
30ec20a7e12a4906bd63aa39143c5639: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '30ec20a7e12a4906bd63aa39143c5639'}
7b8ab9e70ad248138e5647f120cb1d9e: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '7b8ab9e70ad248138e5647f120cb1d9e'}
(29530) accepted ('192.168.1.144', 56831)
9a3a0276f49c4c6b9db86946bb72fd78: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '9a3a0276f49c4c6b9db86946bb72fd78'}
(29530) accepted ('192.168.1.144', 56837)
a4812eaaef844ec9addfef30755c0111: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'a4812eaaef844ec9addfef30755c0111'}
42fbeb1d760d4851bccdd916f4cd791a: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '42fbeb1d760d4851bccdd916f4cd791a'}
```

**Comment by miguelgrinberg** [comment]
Unfortunately this does not clearly point to a server side problem. Clients are making connections, and as far as I can see the server is accepting them, but then nothing happens.

**Comment by egorf** [comment]
Ok, I agree this is vague. However, this not only stops the websocket interactions, but also lives through a page refresh. Only restarting the server helps

Is there a way I can provide more info for you?

**Comment by miguelgrinberg** [comment]
Do you have the entire log? The top line shows that a client timed out, which is odd when you are using websocket. Also, was there any output on stdout?

**Comment by egorf** [comment]
I logged both stdout and stderr using tee. It seems to me that stdout log only contains my app's output(and it is still in development so a lot of garbage messages) but I will still include it. The files are quite long, so you should start from the end.


[stderr.txt](https://github.com/miguelgrinberg/Flask-SocketIO/files/547514/stderr.txt)
[stdout.txt](https://github.com/miguelgrinberg/Flask-SocketIO/files/547515/stdout.txt)

**Comment by egorf** [comment]
A small update...

While waiting for your answers, I've tried different setups including gevent configuration.

After I've installed gevent and monkey patched the app from the very top in my **manage.py**, I've encountered another weird issue: socket interactions(server-side continuous emit in my case) stops immediately after I try downloading a file. 

Is there anything that I'm missing?

**Comment by egorf** [comment]
OK, after seeing that the same thing happens with gevent, I decided to look for atypical code in my app that might have triggered this behaviour.

I've had two things, probably unusual for standard web applications, which caused the issue. Neither of them are direct errors, I think.

- I've had two connect and disconnect handlers in different blueprints
- One of the handlers started a thread, which emits information every second, on connect and stops on disconnect

After removing the connect/disconnect events and starting the thread in the beginning of the app(to emit regardless of the connection availability) server seems to no longer fail, or at least this state is really hard to reach.

I would really like to avoid this endless thread, but otherwise everything seems quite alright. 
Will wait for your comment on this.

**Comment by miguelgrinberg** [comment]
Sorry for the delay in replying. There are a number of things to comment on, I'll try to go one by one.

The stderr logs that you posted show that some of your clients failed to upgrade to websocket, which may be unrelated to your problem, but still interesting. Near the bottom of the log when things start to break, several clients were dropped, which you can see in the "Client is gone" lines. These are long-polling clients that failed to send their pings and have timed out. It's odd that several (maybe all?) long-pollers appear to fail at around the same time, suggesting something lower level isn't working. Were these clients leaving on purpose at that time, or was this unexpected. All these clients then try to reconnect, and that is the end of the log, where you see the server responging with OPEN packages to them.

I'm not sure still what is the problem, unfortunately. You can see that the wsgi layer in eventlet is accepting connections, the server is receiving them and replying with the OPEN packet, so it appears what's missing is that long-polling packets sent by the clients are not arriving to the server. This caused the server to declare these clients gone, as no pings were seen from them, and then when the server sends the open packet in the reconnect there is no reply from them. That's as far as my analysis goes, still not sure what can cause this behavior.

>  I've encountered another weird issue: socket interactions(server-side continuous emit in my case) stops immediately after I try downloading a file.

Not sure why downloading a file would affect other network communications. Was only the client that is downloading affected, or everyone?

> I've figured out you also need gevent-websocket installed for gevent configuration to work

gevent-websocket is not required, but without it all connections are long-polling.

> I've had two connect and disconnect handlers in different blueprints

Flask-SocketIO does not work with blueprints, there is really one handler per event, so setting two means that whichever of the two is the one that is parsed second is the one that sticks.

> One of the handlers started a thread, which emits information every second, on connect and stops on disconnect

This is fine, as long as the thread is not really a thread, but a greenlet. If it is a standard thread, then that's not good. You can use the `socketio.start_background_task()` function to start a background thread that is compatible with your current async framework.

> After removing the connect/disconnect events and starting the thread in the beginning of the app(to emit regardless of the connection availability) server seems to no longer fail, or at least this state is really hard to reach.

Were you running a thread per client before? Sounds like you were, if you started a thread every time you were receiving a connect event. Do you have your code available somewhere I can take a look at?

**Comment by egorf** [comment]
Hello again!

Sorry for the long answer as well! I'll also try to address all your replies.

At this time it's important to know what kind of application I am working on. It is not exactly a website, being more of a web interface for a device. It is not intended be used somewhere other than the local network and not really intended to be used by many clients at once. 

> gevent-websocket is not required, but without it all connections are long-polling.

Well, for me it results in a really weird problem, where the socket interactions stop after clicking a file link, served by a simple flask `@app.route`. Installing **gevent-socketio** helped immediately. I'll try to post some code for you to check in a different issue.

> Was only the client that is downloading affected, or everyone?

I only tried it with a single client as this is my use case most of the time.

> Flask-SocketIO does not work with blueprints

It does, as long as you import the SocketIO instance directly to the blueprints. I really wish I didn't have to, but for a big application using blueprints helps so much.

> there is really one handler per event, so setting two means that whichever of the two is the one that is parsed second is the one that sticks

That's what I was counting on anyway. Just curious if setting two could break something.

> This is fine, as long as the thread is not really a thread, but a greenlet. If it is a standard thread, then that's not good. You can use the `socketio.start_background_task()` function to start a background thread that is compatible with your current async framework.

Well, I use eventlet's `monket_patch()`, so it is a greenlet. Although `socketio.start_background_task()` is really nice to have, it does not offer enough control for me to use.

> Were you running a thread per client before? Sounds like you were, if you started a thread every time you were receiving a connect event

Nope, I just started the thread every time at least somebody connected. 

> Do you have your code available somewhere I can take a look at?

At the moment - not really, sorry.
